src/util/code_observer.h

```diff
@@ -0,0 +1,161 @@
+template<class T> class code_observer_baset
+{
+static_assert(std::is_base_of<codet, T>::value, "T is not derived from codet");
+public:
+  virtual ~code_observer_baset() {}
+
+  /// registers an observer, that is associated with specific codet id.
+  /// \param id: id of the codet to be registered
+  /// \param f: observer callback, that is associated with the given, \p id
+  /// \param args: arguments associated with the given callback, \p f
+  template<class F, class ...Args>
+    void registerObserver(const irep_idt &id, const F &&f, Args&&... args)
+    {
+      auto observer = std::bind(f, args ...);
+      code_observers_map.insert({id, observer});
+    }
+protected:
+  /// notifies (if any) the observer that is registered to receive a callback
+  /// for the given codet, \p code
+  /// \param code: will be passed to the registered observer that
+  /// that has the same associated id
+  void notify(T &code) const
+  {
+    auto it = code_observers_map.find(code.get_statement());
+    if(it != code_observers_map.end())
+      it->second(code);
+  }
+private:
+  using observerst = std::function<void(T&)>;
+  using code_observer_mapt = std::map<const irep_idt, const observerst>;
+  code_observer_mapt code_observers_map;
+};
+
+/// \brief codet observer, uses expression visitor to visit all the codet's
+///   in a given symbol table and appropriately notify registered observers.
+class code_observert final:private expr_visitort,
+  public code_observer_baset<codet>
+{
+public:
+  virtual ~code_observert() {}
+
+  /// iterates through a given symbol table and for every symbol calls it's
+  /// associated expression visitor
+  /// \param symbol_table: symbol table to visit
+  void visit_symbols(symbol_tablet &symbol_table)
+  {
+    forall_symbols(symbol, symbol_table.symbols)
+    {
+      symbolt &w_symbol=symbol_table.get_writeable_ref(symbol->second.name);
+      w_symbol.value.visit(*this);
+    }
+  }
+
+private:
+  /// Called by expr_vistort, for every expression found. Expressions that are
+  /// not of type ID_code are ignored. Otherwise, this function will notify
+  /// the observer (if any) that is associated with this code type.
+  /// \param id: expr: expression
+  void operator()(exprt &expr) override
+  {
+    if(expr.id() != ID_code)
+      return;
+    notify(to_code(expr));
+  };
+};
+
+/// \brief constant version of code_observert
+class const_code_observert final:private const_expr_visitort,
+  public code_observer_baset<const codet>
+{
+public:
+  virtual ~const_code_observert() {}
+
+  /// iterates through a given symbol table and for every symbol calls it's
+  /// associated expression visitor
+  /// \param symbol_table: symbol table to visit
+  void visit_symbols(const symbol_tablet &symbol_table)
+  {
+    forall_symbols(symbol, symbol_table.symbols)
+    {
+      const symbolt &w_symbol=symbol_table.lookup_ref(symbol->second.name);
+      w_symbol.value.visit(*this);
+    }
```

Comment:
As above, except now as the const version:
```
forall_symbols(it, symbol_table.symbols)
{
  it->second.value.visit(*this);
}
```

Reply:
done